### PR TITLE
Create script to generate and upload topics in Dockstore

### PR DESCRIPTION
**Pull request by kathy-t**
**Description**
Webservice PR: https://github.com/dockstore/dockstore/pull/6006

This PR modifies the `generate-topics` command so that it uses the `getAITopicCandidates` endpoint introduced in https://github.com/dockstore/dockstore/pull/6006. 

There are two new parameters to the `generate-topics` command:
- `--max`: This specifies a maximum number of AI topic candidates to retrieve from Dockstore. If not specified, it will retrieve ALL AI topic candidates. We might want to use this for testing, or if we want to be conservative with the number of topics generated.
- `--dryRun`: This retrieves the AI topics from Dockstore and writes the TRS IDs and version IDs to a file. This is similar to the draft PR's `get-topic-candidates` command which no longer exists. Can be combined with the `--max` command.

You can still provide an entries input file if you want to, but by default, the command will get the candidates from Dockstore.

When generating AI topics, the topic generator will write the entries that failed to generate a topic to a file for later inspection or later re-ingestion. 
- Note that with the way that the `getAITopicCandidates` endpoint works, it will return all entries with no topics, even if it had previously failed to have a topic generated (ex: because the descriptor content is missing). Hypothetically, if 100 entries failed, and we requested a max of 100 the next time around and all 100 failed entries were returned again, it would fail again...Maybe a non-problem since we will be requesting all AI topic candidates 

There is also a new script called `generateAndUploadAITopicsToDockstore.sh` that is meant to be used in the nightly topic updates in https://ucsc-cgl.atlassian.net/browse/SEAB-6653. The script invokes the topic generator JAR to generate topics (retrieving the candidates from Dockstore), and uploads them to Dockstore. This script is likely to be updated by Charles' work with the scheduler.

<details>
<summary>Draft PR description</summary>
In draft mode until https://github.com/dockstore/dockstore/pull/6006 is merged. Builds will fail until then.

This PR adds a new `get-topic-candidates` command to the topic generator that gets all the entries from Dockstore that are AI topic candidates and creates a CSV file containing the entries' TRS ID and version name. This CSV file can then be ingested by the `generate-topics` command.

There is also a new script called `updateDockstoreTopics.sh` that is meant to be used in the nightly topic updates in https://ucsc-cgl.atlassian.net/browse/SEAB-6653. The script invokes the topic generator JAR to get the topic candidates, generate topics, and uploads them to Dockstore.

Arguably, the multiple steps could be collapsed into a single command, but this is the first time we're using AI in a feature and this allows us to see the output of each step which simplifies spot checking.
</details>

**Review Instructions**
(pondering how to do this without generating actual AI topics in QA)

**Issue**
[SEAB-6652](https://ucsc-cgl.atlassian.net/browse/SEAB-6652)

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install` in the project that you have modified (until https://ucsc-cgl.atlassian.net/browse/SEAB-5300 adds multi-module support properly)
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If you are changing dependencies, check with dependabot to ensure you are not introducing new high/critical vulnerabilities
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead. 

[SEAB-6652]: https://ucsc-cgl.atlassian.net/browse/SEAB-6652?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: topicgenerator/src/main/java/io/dockstore/topicgenerator/client/cli/TopicGeneratorClient.java [hunk 1]
```diff
@@ -148,7 +179,12 @@ private void generateTopics(TopicGeneratorConfig topicGeneratorConfig, String in
 
                 // Create ChatGPT request
                 try {
-                    getAiGeneratedTopicAndRecordToCsv(openAiService, csvPrinter, trsId, versionId, entryType, descriptorFile);
+                    //getAiGeneratedTopicAndRecordToCsv(openAiService, csvPrinter, trsId, versionId, entryType, descriptorFile);
```
Comment (kathy-t) [inline topicgenerator/src/main/java/io/dockstore/topicgenerator/client/cli/TopicGeneratorClient.java:182]: Will revert this when ready to merge - this is used for testing

**Comment by denis-yuen** [comment]
>  (pondering how to do this without generating actual AI topics in QA)

might not be so bad, I'm hesitant about DOIs because they're immutable and managed by someone else. But topic sentences "just" cost money.  It looks like we're thinking $127 for each time we generate across all workflows in the DB which is not trivial but also not unmanageable?

File: topicgenerator/src/main/java/io/dockstore/topicgenerator/client/cli/TopicGeneratorClient.java [hunk 2]
```diff
@@ -97,16 +102,42 @@ public static void main(String[] args) {
...
+            LOG.info("No AI topic candidates found");
+            return;
+        }
+
+        try (CSVPrinter csvPrinter = new CSVPrinter(new FileWriter(outputCsvFilePath, StandardCharsets.UTF_8), CSVFormat.DEFAULT.builder().setHeader(GenerateTopicsCommand.InputCsvHeaders.class).build())) {
+            aiTopicCandidates.forEach(aiTopicCandidate -> {
+                try {
+                    csvPrinter.printRecord(aiTopicCandidate.getEntryLite().getTrsId(), aiTopicCandidate.getVersionName());
+                } catch (IOException e) {
+                    LOG.error("Could not write record for TRS ID {}, version {}", aiTopicCandidate.getEntryLite().getTrsId(), aiTopicCandidate.getVersionName());
```
Comment (coverbeck) [inline topicgenerator/src/main/java/io/dockstore/topicgenerator/client/cli/TopicGeneratorClient.java:132]: I wonder if it's worth tracking these. I suppose the code I'm suppose to write could capture these in CloudWatch?

File: topicgenerator/scripts/updateDockstoreTopics.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+wget https://artifacts.oicr.on.ca:443/artifactory/collab-release/io/dockstore/topicgenerator/"${TOPIC_GENERATOR_VERSION}"/"${TOPIC_GENERATOR_JAR}"
+
+echo "Creating a CSV file of AI topic candidates from Dockstore"
+java -jar "${TOPIC_GENERATOR_JAR}" -c "${TOPIC_GENERATOR_CONFIG_ABS_PATH}" get-topic-candidates
+
+echo "Generating topics"
+java -jar "${TOPIC_GENERATOR_JAR}" -c "${TOPIC_GENERATOR_CONFIG_ABS_PATH}" generate-topics --entries entries_*.csv
+
+echo "Uploading generated AI topics to Dockstore"
+java -jar "${TOPIC_GENERATOR_JAR}" -c "${TOPIC_GENERATOR_CONFIG_ABS_PATH}" upload-topics --aiTopics generated-topics_*.csv
```
Comment (coverbeck) [inline topicgenerator/scripts/updateDockstoreTopics.sh:35]: This is going to upload these to Dockstore without any vetting, correct? And we're OK with that? I think so, but just double-checking.

**Review by svonworl — APPROVED** [review]

**Comment by kathy-t** [comment]
PR is ready for review, please see the updated PR description. Re-requesting reviews

**Review by svonworl — APPROVED** [review]

**Review by coverbeck — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]

**Review by coverbeck — COMMENTED** [review]
A problem which I just ran into with the scheduled task is `io.dockstore.topicgenerator.client.cli.TopicGeneratorClient#getNextflowMainScript` is calling `io.dockstore.common.NextflowUtilities#grabConfig(java.lang.String)`, which then tries to run Nextflow. And fails because the Nextflow JAR is not on the Docker image.

We have a couple of options:

1. Get the nextflow jar into the Docker image
2. Don't do Nextflow
3. Find another way to get the main script.

**Pull request merged**